kernel/src/drivers/serial.rs

```diff
@@ -0,0 +1,244 @@
+pub static PORT_2: Mutex<SerialPort> = Mutex::new(unsafe { SerialPort::new(PORT_2_ADDR) });
+
+pub struct SerialPort {
+    initialized: bool,
+    data: Port<u8>,
+    interrupt_enable: Port<u8>,
+    fifo_control: Port<u8>,
+    line_control: Port<u8>,
+    modem_control: Port<u8>,
+    line_status: Port<u8>,
+    _modem_status: Port<u8>,
+    _scratch: Port<u8>,
+}
+
+impl SerialPort {
+    pub const unsafe fn new(port_base: u16) -> SerialPort {
+        SerialPort {
+            initialized: false,
+            data: Port::new(port_base),
+            interrupt_enable: Port::new(port_base + 1),
+            fifo_control: Port::new(port_base + 2),
+            line_control: Port::new(port_base + 3),
+            modem_control: Port::new(port_base + 4),
+            line_status: Port::new(port_base + 5),
+            _modem_status: Port::new(port_base + 6),
+            _scratch: Port::new(port_base + 7),
+        }
+    }
+
+    /// Initializes the serial port
+    pub fn init(&mut self, baud: u32, enable_irqs: bool) -> Result<(), InvalidBaudrate> {
+        let divisor = MAX_BAUD / baud;
+        if MAX_BAUD / divisor != baud {
+            return Err(InvalidBaudrate(baud));
+        }
+
+        // Disable interrupts
+        self.interrupt_enable.write(0);
+
+        // Enable DLAB - data port & interrupt enable will temporarily become DLAB lsb & msb
+        self.line_control.write(1 << 7);
+
+        // Write divisor
+        self.data.write((divisor & 0xFF) as u8);
+        self.interrupt_enable.write((divisor >> 8) as u8);
+
+        // 8 bits, no parity byte, one stop bit
+        self.line_control.write(0b111);
+
+        //             Flags: Enable     & Reset tx/rx & 64byte buf & trigger level 56bytes
+        let fifo_flags = (0b1 << 0) | (0b11 << 1) | (0b1 << 5) | (0b11 << 6);
+        self.fifo_control.write(fifo_flags);
+
+        //  Request To Send & Data Terminal Ready
+        let mut modem_ctrl_flags = 0b1 << 0;
+
+        if enable_irqs {
+            modem_ctrl_flags |= 0b1 << 3; // Aux output 2 (enable IRQs, practically)
+        }
+
+        self.modem_control.write(modem_ctrl_flags);
+
+        self.initialized = true;
+        Ok(())
+    }
+
+    /// Returns the line status. Panics if not initialized.
+    pub fn status(&mut self) -> LineStatus {
+        if self.initialized {
+            LineStatus::from_bits_truncate(self.line_status.read())
+        } else {
+            panic!("Serial port not initialized");
+        }
+    }
+
+    /// Attempts to read one byte of data, returning none if no byte was waiting.
+    pub fn try_read(&mut self) -> Option<u8> {
+        if self.status().contains(LineStatus::DATA_READY) {
+            Some(self.data.read())
+        } else {
+            None
+        }
+    }
+
+    /// Attempts to write one byte of data, returning whether it could.
+    pub fn try_write(&mut self, data: u8) -> bool {
```

Comment:
Maybe a `Result<(), ()>`?

Reply:
Eh, i feel like this is a little unnecessary. It's not necesarily an error if it can't -- just means that the receiver is probably busy writing bytes or not ready to accept.

Reply:
Yeah, not too important. Though on second thought a `Result<(), u8>` would work. Either it's successful or it returns the value that it was unable to write

Reply:
Eh.  u8 is copy so there's no reason to, and it would also make it seem like an error code.